distribution/pull_v2.go

```diff
@@ -246,7 +248,20 @@ func (ld *layerDescriptor) Download(ctx context.Context, progressOutput progress
 		ld.verifier = ld.digest.Verifier()
 	}
 
-	_, err = io.Copy(tmpFile, io.TeeReader(reader, ld.verifier))
+	{
+		// Allocate rather large buffers to avoid small
+		// `write()` syscalls to the underlying temporary file.
+		var buf = make([]byte, 524288)
+		var tmpFileBufWriter = bufio.NewWriterSize(tmpFile, 524288)
```

Comment:
512 kiB is a very large buffer -- sixteen times the default 32 kiB buffer size of our `pools.Copy`. How did you pick this value? I have my doubts that the increased memory pressure is worth the 2% speed bost. How much performance would we be leaving on the table if `pools.Copy` was used instead, with the default buffer size?

Reply:
I can try and run some better numbers tomorrow (it's 22 PM, I'm at home and don't have any sort of Go toolchain on this machine, etc. etc.), but my idea was that a layer generally is, say, 20 megs to multiple gigabytes, so a combined megabyte of buffers seems to be pretty peanuts. The size of the buffer could be made a tunable (envvar, maybe?) though?

Without an explicit buffer here, we (empirically) end up doing a lot of pretty small writes; `io.Copy` and `pools.Copy` just use the size of the temporary buffer as the _maximum_ size of a single read, and that single read is then written to the writer - and if that writer is a File, it'll just do the `write` syscall.

Reply:
We have pooled bufio, too.
```suggestion
		var tmpFileBufWriter = pools.BufioWriter32KPool.NewWriteCloserWrapper(pools.BufioWriter32KPool.Get(tmpFile), nil)
		defer tmpFileBufWriter.Close()
```

Reply:
I'll have to dig a little deeper, I think...

I wrote a small standalone program that uses the same basic ideas as `dockerd` here – basically copies bytes from an URL opened with `http.Get()` to a local file with 

* various copy buffer sizes 
* various write buffer sizes
* the `io.Reader` cast/hack workaround from the other comment turned on or off

, and looking at [the results](https://gist.github.com/akx/12b656234e42e6607a0d97c8a761c46f) I'm not really sure I'm measuring the right thing, because it looks like none of this matters in a statistically significant way. (Or I might be measuring the right thing, and none of this actually matters, and it's pretty much as optimum as it can be?)

Reply:
I think in this case the largest read size, regardless of buffer size, is going to be 16KB.
Having a buffered writer could improve things somewhat but perhaps not in a significant way. In this case you are probably only saving the syscall overhead of the write call.